### PR TITLE
[Enhancement] - Replace existing Contact List with MUI Data Grid

### LINKED ISSUE #343: Enhancement: Update Contacts View to display Contact Cards instead of Table
**Describe the Current Behavior/Feature**:
Recent updates have compressed the amount of information displayed in the client's table. It now looks like this:
image

Not very much useful information displayed there.

Our figma mockups also display something else here. Instead of a table, we have a list of cards for each client. We can then click on the client to go to their profile.

**Proposed Behavior/Feature**:
We already have a similar card displayed on the profile page:
image

We could use these cards in place of the table rows to display client information

### PR DESCRIPTION
**Pull request by leekahung**
## This PR:

Updates existing table used for our Contact List with MUI Data Grid. The module for MUI Data Grid has been included to package.json. The column with originally with the link to the contact profile has been split up into the original text and a button icon in order to allow for filtering via MUI Data Grid. Other than that, the grid should function similarly to our previous table.

The unused `ContactListTableRow.jsx` is removed and a new component `ContactProfileIcon.jsx` is created for the React Router Link to the contact profile page.

The PR also includes a minor correction for `contactProfile.profileImage` in `src/components/Profile/ProfileImageField.jsx`. Before the profileImage would be incorrectly rendered because of the incorrect variable name, `profileImg`, which doesn't exist as part of `contactProfile`.

The column with pinning has been removed (UX team believes if there's already a filter + sorting capability with MUI Data Grid, this might not be necessary; it'll be better to have this moved as a type of status instead). The name from Person has been split up into "First Name" and "Last Name" for a better sorting experience.

---
### Components
- `src/components/Contacts/ContactListTable.jsx`
- `src/components/Contacts/ContactListTableRow.jsx` (file removed)
- `src/components/Contacts/ContactProfileIcon.jsx` (file added)
- `src/components/Profile/ProfileImageField.jsx`

### Test Files
- `test/components/Contacts/ContactsListTable.test.jsx`
- `test/components/Profile/ProfileImageField.test.jsx`
- `test/pages/Contacts.test.jsx`

### Other Files
- `src/components/Contacts/index.js`
- `package.json`
- `package-lock.json`

---
## Screenshots (if applicable):

https://github.com/codeforpdx/PASS/assets/14917816/76a2e574-7e2f-488c-8308-866ffff8d7ae

---
## Future Considerations:
- Implement accessibility improvements with MUI Data Grid after merger

### CODE REVIEW

File: test/components/Contacts/ContactsListTable.test.jsx
Comment (milofultz) [inline test/components/Contacts/ContactsListTable.test.jsx]: For this query, you could use `queryByRole` to avoid the `querySelector` call. (see [this and following section in Kent C. Dodds page on avoiding `querySelector`](https://kentcdodds.com/blog/common-mistakes-with-react-testing-library#using-container-to-query-for-elements))

```tsx
  const row1 = queryByRole('cell', { name: 'Aaron Abby' });
  const row2 = queryByRole('cell', { name: 'Bob Builder' });

  expect(row1).not.toBeNull();
  expect(row2).not.toBeNull();
```

File: src/components/Contacts/ContactProfileIcon.jsx
Comment (milofultz) [inline src/components/Contacts/ContactProfileIcon.jsx]: This will render a `button` inside of an `a` tag. Interactive buttons shouldn't be nested because usually one of the two will be ignored. 

From an accessibility perspective, generally [`a` tags](https://www.w3.org/TR/wai-aria-1.2/#link) should be used to navigate and [buttons](https://www.w3.org/TR/wai-aria-1.2/#button) should be used for an action. In this case, I think you could add the `onClick` to the `Link` element without any adverse effects and improve the user experience.

File: src/components/Contacts/ContactListTable.jsx
Comment (milofultz) [inline src/components/Contacts/ContactListTable.jsx:82]: [There are a lot of accessibility issues with this component](https://github.com/mui/mui-x/issues?q=is%3Aopen+label%3A%22component%3A+data+grid%22+label%3Aaccessibility+), but I do think that a `grid` role is the right one so don't want to stand in the way of progress on this point. I tried it with VoiceOver and it wasn't really doing what I was expecting, but don't know if that's our or their side.

**Review by milofultz — CHANGES_REQUESTED** [review]
If we could resolve the button nested in an `a` tag, that would be great. Thank you!

File: src/components/Contacts/ContactProfileIcon.jsx
Comment (leekahung) [reply src/components/Contacts/ContactProfileIcon.jsx]: This is a React Router Link that helps with sending a specific state (i.e. the `contact` object) along with the redirect and browser URL. A simple anchor tag in this case would not do.

File: src/components/Contacts/ContactListTable.jsx
Comment (leekahung) [reply src/components/Contacts/ContactListTable.jsx:82]: Think having a fix for those would be great. We could open up an issue to fix the accessibility problems here in a follow-up. 👍

File: test/components/Contacts/ContactsListTable.test.jsx
Comment (leekahung) [reply test/components/Contacts/ContactsListTable.test.jsx]: I'll give this a try.

**Comment by leekahung** [comment]
> If we could resolve the button nested in an `a` tag, that would be great. Thank you!

Thanks for the review @milofultz! I'll have the changes made in a bit.

File: src/components/Contacts/ContactProfileIcon.jsx
Comment (leekahung) [reply src/components/Contacts/ContactProfileIcon.jsx]: Or do you mean to just move the onClick to Link and just leave the Icon wrapped without the IconButton? I could try that.

**Comment by andycwilliams** [comment]
Is there an associated issue for this?

@milofultz This is valuable feedback, at least for my own understanding. More insight into accessibility is really helpful for the project.

**Comment by leekahung** [comment]
> Is there an associated issue for this? 

I was looking around the issues. There doesn't seem to be one for this or the ones that were there was closed.

File: test/components/Contacts/ContactsListTable.test.jsx
Comment (leekahung) [reply test/components/Contacts/ContactsListTable.test.jsx]: Got this updated in fix.

File: src/components/Contacts/ContactProfileIcon.jsx
Comment (leekahung) [reply src/components/Contacts/ContactProfileIcon.jsx]: Got this changed to:

```jsx
    <Link
      to={`/profile/${encodeURIComponent(contact.id)}`}
      state={{ contact: contact.value }}
      onClick={() => handleSelectProfile(contact.value)}
    >
      <ContactPageIcon sx={{ color: 'gray' }} titleAccess="contact profile link" />
    </Link>
```

**Comment by timbot1789** [comment]
I think the data grid is a good enhancement over the current view. I especially like how we can easily add sorting to it (which may eliminate the need for pinning / favoriting). 

How would you say it relates to [this ticket](https://github.com/codeforpdx/PASS/issues/343), where I talk about converting the contacts list into a series of cards?
I based that off of [one of the figma designs](https://www.figma.com/file/BfsGjGbFUVvtqNq18sykoY/Pass-Project?type=design&node-id=1216-2986&mode=design) UX submitted. 
image

**Comment by leekahung** [comment]
> How would you say it relates to [this ticket](https://github.com/codeforpdx/PASS/issues/343), where I talk about converting the contacts list into a series of cards? I based that off of [one of the figma designs](https://www.figma.com/file/BfsGjGbFUVvtqNq18sykoY/Pass-Project?type=design&node-id=1216-2986&mode=design) UX submitted.

I think we could certainly have this link this to that issue since it does update the UI.

If we want to, think we could also separate out different fields from `contact` (like first name, last name) in a follow-up PR to make better use of Data Grid's innate sorting capabilities instead of our existing sorting scheme.

**Comment by andycwilliams** [comment]
> > How would you say it relates to [this ticket](https://github.com/codeforpdx/PASS/issues/343), where I talk about converting the contacts list into a series of cards? I based that off of [one of the figma designs](https://www.figma.com/file/BfsGjGbFUVvtqNq18sykoY/Pass-Project?type=design&node-id=1216-2986&mode=design) UX submitted.
> 
> I think we could certainly have this link this to that issue since it does update the UI.
> 
> If we want to, think we could also separate out different fields from `contact` (like first name, last name) in a follow-up PR to make better use of Data Grid's innate sorting capabilities instead of our existing sorting scheme.

When I last talked to UX about this, they said they were still working on the desktop version of these cards. So that combined with temporarily focusing on documentation recently, I haven't done a ton with #343 lately (and should really have moved that to the backlog).

I'm still willing to complete that, it would just mean waiting until this is merged before continuing.

Also, I would likely focus on #414 before moving onto #343.

**Comment by timbot1789** [comment]
> When I last talked to UX about this, they said they were still working on the desktop version of these cards. So that combined with temporarily focusing on documentation recently, I haven't done a ton with #343 lately (and should really have moved that to the backlog).
> 
> I'm still willing to complete that, it would just mean waiting until this is merged before continuing.
> 
> Also, I would likely focus on #414 before moving onto #343.

That sounds good to me. I personally like the card look more than the table. IMO it's more humanizing, and shows the data in a more familiar way. And I think it will work well in a desktop environment. The table view is good when you have several rows worth of data you need to look at all at once, but I don't know how often that will be the case here.

File: src/components/Contacts/ContactListTable.jsx
Comment (xscottxbrownx) [inline src/components/Contacts/ContactListTable.jsx]: should the contact name not be sortable?

Comment (xscottxbrownx) [reply src/components/Contacts/ContactListTable.jsx]: I assume this is the arrow that shows if you hover and makes sorted by ascending or descending values...

Comment (xscottxbrownx) [reply src/components/Contacts/ContactListTable.jsx]: If so, personally I would think this to be one of the most used features of the grid honestly.

File: package.json
Comment (andycwilliams) [inline package.json:37]: Isn't this data grid what we used months back when creating the original table and implementing MUI? I don't recall why we decided not to go with it at that time.

File: src/components/Contacts/ContactProfileIcon.jsx
Comment (milofultz) [reply src/components/Contacts/ContactProfileIcon.jsx]: Yeah, that's what I was thinking. Thank you!

**Comment by milofultz** [comment]
> > Is there an associated issue for this?
> 
> I was looking around the issues. There doesn't seem to be one for this or the ones that were there was closed.

@andycwilliams @leekahung The issues I had when using VoiceOver on the data grid (which I'll admit isn't technically super proficient) were just basic navigation. In comparison to [a data grid like this one from W3](https://www.w3.org/WAI/ARIA/apg/patterns/grid/examples/data-grids/#aboutthisexample) which was super easy to navigate and interact with, the MUI one on their site was almost unusable. 

I don't quite have to technical know how with grid yet to understand why this is the case, but I saw similar complaints in their backlog (https://github.com/mui/mui-x/issues/4749, https://github.com/mui/mui-x/issues/4658, https://github.com/mui/mui-x/issues/4282).

Again, don't know how much of this is them, us, or me, but I saw a little bit of signal confirming my suspicions in their backlog so thought I'd bring it up.

edit: oop misread the original post, but leaving this here for clarity anyway, sorry for the erroneous tag now :)

File: src/components/Contacts/ContactListTable.jsx
Comment (leekahung) [reply src/components/Contacts/ContactListTable.jsx]: I was thinking about sorting with what we have now. Was thinking we could perhaps split up `person` back into just `givenName` and `familyName`. If we do that, think having the sorting then would be better. See clip:

https://github.com/codeforpdx/PASS/assets/14917816/c749ec9e-218e-4895-9d95-eea30d59d8a3

It'll open up more fields for filtering as well.

https://github.com/codeforpdx/PASS/assets/14917816/e1df46fd-f968-4d14-942f-a180a19925e6

The changes done in those samples are relatively simple.

File: package.json
Comment (leekahung) [reply package.json:37]: Yeah, it was the same one as before. I think at the time we didn't went with it was because there were still a few thing we wish to discuss, although I don't remember which things.

File: src/components/Contacts/ContactProfileIcon.jsx
Comment (milofultz) [inline src/components/Contacts/ContactProfileIcon.jsx:49]: thanks again! 🙏 🙇

File: test/components/Contacts/ContactsListTable.test.jsx
Comment (milofultz) [inline test/components/Contacts/ContactsListTable.test.jsx]: could use the same `getByRole` schema here as above

**Review by milofultz — APPROVED** [review]
Just one small thing, but not a blocker

File: test/components/Contacts/ContactsListTable.test.jsx
Comment (leekahung) [reply test/components/Contacts/ContactsListTable.test.jsx]: No problem! I'll have this switched up before merging.

**Comment by leekahung** [comment]
Got everything updated from the suggestions. Let me know if there's still anything that needs to be reviewed for this PR.

File: test/components/Contacts/ContactsListTable.test.jsx
Comment (leekahung) [reply test/components/Contacts/ContactsListTable.test.jsx]: Got this updated to use `getByRole` 👍

File: src/components/Contacts/ContactListTable.jsx
Comment (leekahung) [reply src/components/Contacts/ContactListTable.jsx]: Got Person split up into First Name and Last Name and allow sorting of those fields. The clip for the update is now shown in the PR description. Let me know if you think this is good.

**Comment by leekahung** [comment]
Hey @xscottxbrownx @andycwilliams. Think I've manage to touch on the things you guys mentioned. Let me know if there's still other things you guys need to review or it's good to go?

File: src/components/Contacts/ContactListTable.jsx
Comment (andycwilliams) [inline src/components/Contacts/ContactListTable.jsx]: With the new data grid, is a custom sorting function still necessary? Unless I'm mistaken, it should be possible to use `initialState` in `DataGrid` for the same results.

File: src/components/Contacts/ContactListTable.jsx
Comment (andycwilliams) [inline src/components/Contacts/ContactListTable.jsx]: Could we possibly expand the table a bit when in desktop views? The quickest way would be by widening the columns.

I know there isn't a ton of data to display at the moment, but it'd help to use up some of all the empty space. Plus, in case anyone has a longer name, it'd prevent that from being cut off.

File: src/components/Contacts/ContactListTable.jsx
Comment (andycwilliams) [inline src/components/Contacts/ContactListTable.jsx:70]: Although it's not essential, I think it'd be nicer to have a title for the delete column. If only because it's a bit odd to have one nameless column. The same kind of column in the documents is named, so this'd make it more consistent.

Adding `headerName: 'Delete',` here would accomplish this.

**Review by andycwilliams — COMMENTED** [review]
Overall, it looks really nice. It's strange seeing the data grid again after all these months.

I think it would have been good to first create an issue before opening a PR as we normally do.

Curious, is the grid also intended to be implemented with the documents table?

**Comment by leekahung** [comment]
> Curious, is the grid also intended to be implemented with the documents table?

Yeah, I think the grid would work well for the document list as well. Think we could have an issue created for that.

File: src/components/Contacts/ContactListTable.jsx
Comment (leekahung) [reply src/components/Contacts/ContactListTable.jsx]: We could certainly drop the custom sort and leave it as an initial state.

File: src/components/Contacts/ContactListTable.jsx
Comment (leekahung) [reply src/components/Contacts/ContactListTable.jsx:70]: I'll add one. I'll expand the width of this too.

File: src/components/Contacts/ContactListTable.jsx
Comment (leekahung) [reply src/components/Contacts/ContactListTable.jsx]: Would like to keep the minimum width of the overall grid to 300px for mobile, but I'll play around with the column widths.

File: src/components/Contacts/ContactListTable.jsx
Comment (leekahung) [reply src/components/Contacts/ContactListTable.jsx:70]: Got this included.

File: src/components/Contacts/ContactListTable.jsx
Comment (leekahung) [reply src/components/Contacts/ContactListTable.jsx]: Made text columns flexible. If we get more columns included in between them in the future, we could make them flexible as well.

https://github.com/codeforpdx/PASS/assets/14917816/1ab60850-c076-4f2b-95d3-7877a2309e0b

File: src/components/Contacts/ContactListTable.jsx
Comment (leekahung) [reply src/components/Contacts/ContactListTable.jsx]: Got the custom sort dropped for sorting from `initialState`

**Comment by xscottxbrownx** [comment]
Issue with rendering size on skinny mobile devices (like Galaxy Fold and similar), but not too worried about for MVP.

**Comment by leekahung** [comment]
> Issue with rendering size on skinny mobile devices (like Galaxy Fold and similar), but not too worried about for MVP.

Yeah, I believe those screen sizes are at around 280px width? At an earlier discussion, I believe we've talked about limiting the lower screen limit to about 300px. Think I could adjust it to make it fit the smaller size using `vw` for the width size on the box surrounding the grid.

https://github.com/codeforpdx/PASS/assets/14917816/7ccaa6fb-1468-4e81-b17b-7301bc07353b

**Review by andycwilliams — COMMENTED** [review]
Looking better. There's a plan to add at least one more column (an icon button to allow for direct messaging), so that will help fill out the table.

I'm seeing some large gaps between the navbar and table (see screenshot) when using larger viewports. Looks to be caused by `justifyContent: 'center'` on line 58 in `Contacts.jsx`

Capture

Is this the desired effect? It might be nicer to have it consistent on all viewports and make the table always close to the navbar. Just want to be sure.

I'd really like to hear from case workers about how small the screens they actually use are. It may not even be an issue to set a 300px limit. But it may take time to get feedback.

**Comment by leekahung** [comment]
> I'm seeing some large gaps between the navbar and table (see screenshot) when using larger viewports. Looks to be caused by `justifyContent: 'center'` on line 58 in `Contacts.jsx`

I could certainly pull the Data Grid back up closer the the Navbar without the `justifyContent`. I'll also make the grid use a set height to fill out the the spacing in the "Contacts" page. It would not affect the mobile view.

https://github.com/codeforpdx/PASS/assets/14917816/82e3a4d7-ebd7-4be2-ab09-c50660de02ca

> I'd really like to hear from case workers about how small the screens they actually use are. It may not even be an issue to set a 300px limit. But it may take time to get feedback.

With regards to the mobile width limit, what we're making here should be quite flexible down to 280px and lower, so I don't think we need to worry about that point now. (As you can see below, the grid content stretches out to 380px, but the grid itself can remain small (around ~210px in the example clip) thanks to the horizontal scrolling implemented with the grid.)

https://github.com/codeforpdx/PASS/assets/14917816/f6f04cfc-fb9d-46b5-bb66-75efae38edea

**Comment by leekahung** [comment]
I believe the changes here are now adequate?

**Pull request merged**
